cpp/src/transform/cudf_to_arrow.cpp

```diff
@@ -0,0 +1,180 @@
+{
+  const int64_t data_size_in_bytes = sizeof(T) * input_view.size();
+  std::shared_ptr<arrow::Buffer> data_buffer;
+
+  CUDF_EXPECTS(arrow::AllocateBuffer(ar_mr, data_size_in_bytes, &data_buffer).ok(),
+               "Failed to allocate Arrow buffer for data");
+  cudaMemcpy(
+    data_buffer->mutable_data(), input_view.data<T>(), data_size_in_bytes, cudaMemcpyDeviceToHost);
+
+  return data_buffer;
+}
+
+std::shared_ptr<arrow::Buffer> fetch_mask_buffer(column_view input_view, arrow::MemoryPool* ar_mr)
+{
+  const int64_t mask_size_in_bytes = cudf::bitmask_allocation_size_bytes(input_view.size());
+  std::shared_ptr<arrow::Buffer> mask_buffer;
+
+  if (input_view.has_nulls()) {
+    CUDF_EXPECTS(arrow::AllocateBuffer(ar_mr, mask_size_in_bytes, &mask_buffer).ok(),
+                 "Failed to allocate Arrow buffer for mask");
+    if (input_view.offset() > 0) {
+      cudaMemcpy(mask_buffer->mutable_data(),
+                 cudf::copy_bitmask(input_view).data(),
+                 mask_size_in_bytes,
+                 cudaMemcpyDeviceToHost);
+    } else {
+      cudaMemcpy(mask_buffer->mutable_data(),
+                 input_view.null_mask(),
+                 mask_size_in_bytes,
+                 cudaMemcpyDeviceToHost);
+    }
+
+    return mask_buffer;
+  }
+
+  return nullptr;
+}
+
+struct dispatch_to_arrow {
+  template <typename T>
+  std::enable_if_t<is_fixed_width<T>(), std::shared_ptr<arrow::Array>> operator()(
+    column_view input_view, cudf::type_id id, arrow::MemoryPool* ar_mr)
+  {
+    auto data_buffer = fetch_data_buffer<T>(input_view, ar_mr);
+    auto mask_buffer = fetch_mask_buffer(input_view, ar_mr);
+
+    return to_arrow_array(id,
+                          static_cast<int64_t>(input_view.size()),
+                          data_buffer,
+                          mask_buffer,
+                          static_cast<int64_t>(input_view.null_count()));
+  }
+
+  template <typename T>
+  std::enable_if_t<is_compound<T>(), std::shared_ptr<arrow::Array>> operator()(
+    column_view input, cudf::type_id id, arrow::MemoryPool* ar_mr)
+  {
+    std::unique_ptr<column> tmp_column = nullptr;
+    if ((input.offset() != 0) or
+        (std::is_same<T, cudf::dictionary32>::value and input.child(0).size() != input.size()) or
+        (!std::is_same<T, cudf::dictionary32>::value and
+         (input.child(0).size() - 1 != input.size()))) {
+      tmp_column = std::make_unique<cudf::column>(input);
+    }
+
+    column_view input_view = (tmp_column != nullptr) ? tmp_column->view() : input;
+
+    auto mask_buffer = fetch_mask_buffer(input_view, ar_mr);
+    std::vector<std::shared_ptr<arrow::Array>> child_arrays;
+
+    for (size_type i = 0; i < input_view.num_children(); i++) {
+      auto c = input_view.child(i);
+      child_arrays.emplace_back(
+        type_dispatcher(c.type(), dispatch_to_arrow{}, c, c.type().id(), ar_mr));
+    }
+
+    if (std::is_same<T, cudf::string_view>::value) {
```

Comment:
Instead of putting all of the logic for all of the compound types in the same `is_compound` path, I think it would be much cleaner to have separate specializations for the different types.